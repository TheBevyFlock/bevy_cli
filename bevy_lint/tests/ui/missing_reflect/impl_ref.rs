//! Tests the `missing_reflect` lint when `Component`, `Resource`, and `Event` are manually
//! implemented for references.
//!
//! This test ensures that the lint calls `peel_refs()` on the type in order to correctly locate
//! the source definition.
//!
//! Note that all 3 traits require `'static` bounds, so generic `'a` lifetimes are not tested.

#![feature(register_tool)]
#![register_tool(bevy)]
#![allow(dead_code)]
//~v NOTE: the lint level is defined here
#![deny(bevy::missing_reflect)]

use bevy::{
    ecs::component::{Mutable, StorageType},
    prelude::*,
};

//~| HELP: `Reflect` can be automatically derived
//~v ERROR: defined a component without a `Reflect` implementation
struct MyComponent;

//~v NOTE: `Component` implemented here
impl Component for &'static MyComponent {
    const STORAGE_TYPE: StorageType = StorageType::Table;
    type Mutability = Mutable;
}

//~| HELP: `Reflect` can be automatically derived
//~v ERROR: defined a resource without a `Reflect` implementation
struct MyResource {
    field_1: usize,
    field_2: bool,
}

//~v NOTE: `Resource` implemented here
impl Resource for &'static &'static MyResource {}

//~| HELP: `Reflect` can be automatically derived
//~v ERROR: defined an event without a `Reflect` implementation
struct MyEvent(String);

impl Component for &'static &'static &'static MyEvent {
    const STORAGE_TYPE: StorageType = StorageType::Table;
    type Mutability = Mutable;
}

//~v NOTE: `Event` implemented here
impl Event for &'static &'static &'static MyEvent {
    type Traversal = ();
}
