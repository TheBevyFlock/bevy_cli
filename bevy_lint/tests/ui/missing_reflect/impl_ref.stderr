error: defined an event without a `Reflect` implementation
  --> tests/ui/missing_reflect/impl_ref.rs:38:1
   |
38 | struct MyEvent(String);
   | ^^^^^^^^^^^^^^^^^^^^^^^
   |
note: `Event` implemented here
  --> tests/ui/missing_reflect/impl_ref.rs:45:1
   |
45 | impl Event for &'static &'static &'static MyEvent {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: the lint level is defined here
  --> tests/ui/missing_reflect/impl_ref.rs:13:9
   |
13 | #![deny(bevy::missing_reflect)]
   |         ^^^^^^^^^^^^^^^^^^^^^
help: `Reflect` can be automatically derived
   |
38 + #[derive(Reflect)]
39 | struct MyEvent(String);
   |

error: defined a component without a `Reflect` implementation
  --> tests/ui/missing_reflect/impl_ref.rs:19:1
   |
19 | struct MyComponent;
   | ^^^^^^^^^^^^^^^^^^^
   |
note: `Component` implemented here
  --> tests/ui/missing_reflect/impl_ref.rs:22:1
   |
22 | / impl Component for &'static MyComponent {
23 | |     const STORAGE_TYPE: StorageType = StorageType::Table;
24 | | }
   | |_^
help: `Reflect` can be automatically derived
   |
19 + #[derive(Reflect)]
20 | struct MyComponent;
   |

error: defined a resource without a `Reflect` implementation
  --> tests/ui/missing_reflect/impl_ref.rs:28:1
   |
28 | / struct MyResource {
29 | |     field_1: usize,
30 | |     field_2: bool,
31 | | }
   | |_^
   |
note: `Resource` implemented here
  --> tests/ui/missing_reflect/impl_ref.rs:34:1
   |
34 | impl Resource for &'static &'static MyResource {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: `Reflect` can be automatically derived
   |
28 + #[derive(Reflect)]
29 | struct MyResource {
   |

error: aborting due to 3 previous errors

